src/compiler/checker.ts

```diff
@@ -3643,9 +3630,17 @@ module ts {
         }
 
         function getReducedTypeOfUnionType(type: UnionType): Type {
-            // If union type was created without subtype reduction, perform the deferred reduction now
+            // If union type was created without subtype reduction, perform the deferred reduction now. If a circularity
+            // is detected, simply use the type itself.
             if (!type.reducedType) {
-                type.reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                type.reducedType = resolvingType;
+                let reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                if (type.reducedType === resolvingType) {
+                    type.reducedType = reducedType;
+                }
+            }
+            else if (type.reducedType === resolvingType) {
+                type.reducedType = type;

```

Comment:
That makes sense.  We should definitely state this in the code itself with a comment of some sort.  It's not at all evident, and it's definitely important for the code to make this clear that this is expected and desirable behavior.